### DISCUSSION TITLE
Q: How to single query insert into unique constraint table

### DISCUSSION THREAD

**Discussion by Fractal-Tess** [prisma/prisma#14665] [Q&A] [answered]
Hey there,

So I have a question about inserting into a table with constraints on some columns.
Let's take a `user` table for example: It should have an `id - auto increment primary key and so on`, `email(unique - string)`, `username(unique - string)`, `password(string)`.

So my goal would be to write to the table with a single query, in case the new user has a unique `username` and `email` within the table, I would like to create a new record, otherwise, fail the insertion and construct an error object.
Now, all of this is already built into prisma, we have the code === p2002 check, and we can access the message on the error object via the `.message` property, however it is in an awful format like `Unique constraint failed on the fields: ('username')`. Also, I found out that even if both the `email` and `username` we are trying to create already exist, only one of them would show up in the error message - so I suppose it throws on the first one that fails to pass the unique test and doesn't check any further.

My approach to this was to just embrace the two query approach and first check for unique duplicates. Here is the code
```ts
const user = await prisma.user.findMany({
    where: {
      OR: [{ username: reg.username }, { email: reg.email }]
    }
  });
  if (user.length > 0) {
    const email = user.map((u) => u.email === reg.email).includes(true);
    const username = user
      .map((u) => u.username === reg.username)
      .includes(true);

    console.log(
      `Already in database are Email:${email}, and username:${username}`
    );
  // Handle constructing an error
  return;
  }
// Handle insertion into the database as there are no duplicates
```

So, my question boils down to: `Is this wrong?`, `Can it be Improved and how?`

Thank you for your time,
Fractal-Tess

**Comment by austincrim** [comment, marked as answer]
Hey @Fractal-Tess 👋,

I apologize for the delay in getting to this question. Let me know if this is still an issue for you.

I don't see any problems with your solution, though I would wrap this logic in an [interactive transaction](https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide#interactive-transactions-in-preview).

**Reply by Fractal-Tess** [reply]
Hey there @austincrim,

Thank you for your reply, regardless of the delay.
As far as the question is concerned, I feel I did not express what I was trying to achieve, so here we go again.

From this question, I would like to receive some clarification on the intended way to write into a table with unique constraints.
I shall provide an example: 
We have table `user` with 3 columns, 2 of which are unique.

```bash
email - unique
username - unique
password - no restrictions
```

Suppose I need to create a new user in this table - What is the optimal way of doing so?
In my original question, I showed how I was first making a query into the database to see if there are users that exist with such credentials with this code 
```ts
const users = prisma.user.findMany({
    where: {
      OR: [{ username: reg.username }, { email: reg.email }]
    }
  });
  ```
So the logic here would be if the returned array from `findMany` is empty (length of 0), then that means the query for the email and username have *not* been already taken by another user, so it is safe to insert a new user with these credentials them into the table.
To me, that doesn't seem like an optimal solution, because if someone tries to register 2 times with the same credentials at the same point in time, there may be a data race and I'm pretty confident that one of the insert statements would fail, causing the application to throw.

How exactly should this functionality be implemented the correct way?

**Reply by austincrim** [reply]
I think this is the correct way to implement it. If you have a data race, one of the writes is going to fail no matter what, and you will need to handle that error in your app code if you want to avoid a crash.

Another option is to use [`$executeRaw`](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#executeraw) with an `ON CONFLICT` clause.

**Reply by Fractal-Tess** [reply]
My general idea was to make it into a single query, where I would insert the new user into the database without first checking if the email or username is already in use. If neither of them is in use, the record creation should be successful and everything should resolve without a hitch, however, if one of them or both of them are already taken, I would've liked the insert function to either return me an error object or throw with some useful information.

The only complaint I have with Prisma is the error object on tables with unique constraints is hard to consume and pinpoint the exact cause of the error.

I think this has answered my question, thank you for your time!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
